### PR TITLE
Shared Network Firewall (Security groups) in Advanced zone without security groups

### PR DESCRIPTION
**Pull request by vishesh92**
Doc PR: https://github.com/apache/cloudstack-documentation/pull/422
### Description

This PR allows creating shared networks with Security Groups in a Zone with securitygroupenabled set to false. This flag is used to prevent creation of isolated networks as of now to maintain backward compatibility for Zone with Security Groups.

To enable SG, enable the SecurityGroupProvider for Guest physical network. After this change, you will be able to create a shared network with Security Groups.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test alma9 kvm-alma9

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
> @vishesh92 will you move the integration tests to `smoke` directory ?

These tests take too long to run (1-2 hours). It's better to skip them.

**Comment by vishesh92** [comment]
@blueorangutan test keepEnv

**Comment by vishesh92** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -6165,6 +6165,28 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
...
+                Network network = _networkDao.findById(networkId);
+                if (network == null) {
+                    throw new InvalidParameterValueException("Unable to find network by id " + networkId);
+                }
+                if (network.getGuestType() == Network.GuestType.Shared && _networkModel.isSecurityGroupSupportedInNetwork(network)) {
+                    return true;
+                }
+            }
+        } else if (CollectionUtils.isNotEmpty(securityGroupsIds)) {
+            Network networkWithSecurityGroup = _networkModel.getNetworkWithSGWithFreeIPs(zone.getId());
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6184]: @vishesh92 
can you explain why added this ?

File: test/integration/component/test_advancedsg_networks.py [hunk 2]
```diff
@@ -169,6 +171,26 @@ def setUpClass(cls):
         # Get Zone, Domain and templates
         cls.domain = get_domain(cls.api_client)
         cls.zone = get_zone(cls.api_client, cls.testClient.getZoneForTests())
+        if cls.zone.securitygroupsenabled is False:
```
Comment (weizhouapache) [inline test/integration/component/test_advancedsg_networks.py:174]: component test won't run in trillian test

can you move to smoke ? if some tests are not really necessary, we can remove `"advanced"` from `@attr(tags`
if everything is fine, we can move back to component if it takes long time.

File: ui/src/views/infra/network/ServiceProvidersTab.vue [hunk 3]
```diff
@@ -937,10 +937,38 @@ export default {
             }
           ]
         },
-        // {
-        //   title: 'SecurityGroupProvider',
-        //   details: ['name', 'state', 'id', 'servicelist'],
-        // },
+        {
+          title: 'SecurityGroupProvider',
```
Comment (weizhouapache) [inline ui/src/views/infra/network/ServiceProvidersTab.vue:941]: is it disabled by default in advanced zone ?
can you add a doc PR to tell how to support SG ? it seems users only need to enable the provider ?

can SG be supported per network or per account ?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 1 above)
Comment (vishesh92) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6184]: If security group IDs are not empty, then we use a default security group in `createAdvancedSecurityGroupVirtualMachine`.

File: test/integration/component/test_advancedsg_networks.py (hunk 2 above)
Comment (vishesh92) [reply test/integration/component/test_advancedsg_networks.py:174]: As of now these tests take around 20 minutes. Let me see what all tests we can disable and move this file to smoke tests.

File: ui/src/views/infra/network/ServiceProvidersTab.vue (hunk 3 above)
Comment (vishesh92) [reply ui/src/views/infra/network/ServiceProvidersTab.vue:941]: Let me create the docs for this PR

File: test/integration/component/test_advancedsg_networks.py (hunk 2 above)
Comment (vishesh92) [reply test/integration/component/test_advancedsg_networks.py:176]: @weizhouapache I ran the acceptance tests. And each tests seems to be taking around 3-5 minutes. All of these tests in total are taking around 2 hours. What all tests do you think we need to keep in smoke?

Class | Test Name | Time
--- | --- | ---
TestNetworksInAdvancedSG | test_03_createIsolatedNetwork | 211.864
TestNetworksInAdvancedSG | test_05_deployVM_SharedwithSG | 316.476
TestNetworksInAdvancedSG | test_06_SharedNwSGAccountSpecific | 218.468
TestNetworksInAdvancedSG | test_07_SharedNwSG_DomainWide_SubdomainAcccess | 306.024
TestNetworksInAdvancedSG | test_10_deleteSharedNwSGAccountSpecific_InUse | 314.851
TestNetworksInAdvancedSG | test_11_deleteSharedNwSG_DomainWide_InUse | 303.817
TestNetworksInAdvancedSG | test_12_deleteSharedNwSGAccountSpecific_NotInUse | 215.508
TestNetworksInAdvancedSG | test_13_deleteSharedNwSG_DomainWide_NotInUse | 212.807
TestNetworksInAdvancedSG | test_29_deleteSharedNwSG_ZoneWide_InUse | 305.857
TestNetworksInAdvancedSG | test_30_deleteSharedNwSG_ZoneWide_NotInUse | 211.688
TestNetworksInAdvancedSG | test**14_createSharedNwWithSG_withoutParams | 214.441
TestNetworksInAdvancedSG | test**15_createVPC | 264.207
TestNetworksInAdvancedSG_VmOperations | test_18_DeployVM_NoFreeIPs_1_account | 308.510
TestNetworksInAdvancedSG_VmOperations | test_18_DeployVM_NoFreeIPs_2_domain | 310.563
TestNetworksInAdvancedSG_VmOperations | test_20_DeployVM_SecGrp_sharedNetwork_1_default | 308.611
TestNetworksInAdvancedSG_VmOperations | test_20_DeployVM_SecGrp_sharedNetwork_2_custom | 305.574
TestNetworksInAdvancedSG_VmOperations | test_24_DeployVM_Multiple_Shared_Networks | 353.151
TestNetworksInAdvancedSG_VmOperations | test_25_Deploy_Multiple_VM_Different_Shared_Networks_Same_SG | 396.687
TestNetworksInAdvancedSG_VmOperations | test_26_Destroy_Deploy_VM_NoFreeIPs | 474.784
TestNetworksInAdvancedSG_VmOperations | test_31_Deploy_VM_multiple_shared_networks_sg | 355.199
TestNetworksInAdvancedSG_VmOperations | test_33_VM_Migrate_SharedNwSG_1_account | 0.000
TestNetworksInAdvancedSG_VmOperations | test**16_AccountSpecificNwAccess | 408.727
TestNetworksInAdvancedSG_VmOperations | test**17_DomainSpecificNwAccess | 400.732
TestSecurityGroups_BasicSanity | test_32_delete_default_security_group | 214.162
TestSharedNetworkOperations | test_34_restart_shared_network_sg_1_account | 261.479
TestSharedNetworkOperations | test_34_restart_shared_network_sg_2_domain | 258.813
TestSharedNetworkOperations | test_34_restart_shared_network_sg_3_zone | 258.858

**Review by weizhouapache — APPROVED** [review]
overall code lgtm

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 1 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6184]: got it

as discussed, if securityGroupsIds is specified but networkids are not, ACS will choose a shared network with SG to deploy the vm

File: test/integration/component/test_advancedsg_networks.py (hunk 2 above)
Comment (weizhouapache) [reply test/integration/component/test_advancedsg_networks.py:176]: ok, let's keep as component tests
it is good you have run the tests and all passed @vishesh92

File: ui/src/store/modules/user.js [hunk 4]
```diff
@@ -342,6 +349,15 @@ const user = {
           reject(error)
         })
 
+        api(
+          'listNetworklistNetworkServiceProviders',
```
Comment (vishesh92) [inline ui/src/store/modules/user.js:353]:
```suggestion
          'listNetworkServiceProviders',
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
clgtm, let's run the relevant component tests before merge

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan test keepEnv

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
clgtm, needs testing

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Review by kiranchavala — APPROVED** [review]
LGTM, 

Tested the security groups for shared network on a Cloudstack advanced zone +kvm

The SG Rules are applied successfully to the vm's and I was also able to change the security groups associated to the vm from the UI

**Pull request merged**